sbi/samplers/vi/vi_divergence_optimizers.py

```diff
@@ -0,0 +1,687 @@
+from __future__ import annotations
+
+import torch
+from torch.distributions import Distribution
+from torch import Tensor
+from torch import nn
+from torch.optim import Adam, Optimizer
+from torch.optim.lr_scheduler import ExponentialLR
+
+from typing import Optional, List, Tuple, Callable, Dict
+from sbi.types import Array
+
+
+import numpy as np
+from abc import ABC, abstractmethod
+from copy import deepcopy
+
+from sbi.inference.potentials.base_potential import BasePotential
+
+
+from sbi.samplers.vi.vi_utils import (
+    filter_kwrags_for_func,
+    make_object_deepcopy_compatible,
+    move_all_tensor_to_device,
+)
+
+
+_VI_method = {}
+
+
+class DivergenceOptimizer(ABC):
+    """This is a wrapper round a PyTorch optimizer and scheduler, which will be used to
+    learn the variational distribution. It further contains some methods to evaluate
+    convergence and to recover a valid state if something wents wrong.
+
+    This class contains one abstract method '_generate_loss_function' which must be
+    implemented und determines the loss function used within variational inference."""
+
+    def __init__(
+        self,
+        potential_fn: BasePotential,
+        q: Distribution,
+        n_particles: int = 256,
+        clip_value: float = 5.0,
+        optimizer: "Optimizer" = Adam,
+        scheduler: object = ExponentialLR,
+        eps: float = 1e-5,
+        **kwargs,
+    ):
+        """This is a wrapper around a PyTorch optimizer which is used to minimize some
+         loss for variational inference.
+
+        Args:
+            potential_fn: Potential function of the target i.e. the posterior density up
+                to normalization constant.
+            q: Variational distribution
+            n_particles: Number of samples used to estimate gradients.
+            clip_value: Norm value on which gradients are clipped.
+            optimizer: Base class for an pytorch optimizer.
+            scheduler: Base class for an pytorch scheduler.
+            eps: This value determines the sensitivity of the convergence checks.
+            kwargs: All additional arguments associated with optimizer, scheduler such
+                as learning_rate.
+
+
+        """
+
+        self.potential_fn = potential_fn
+        self.q = q
+        self.prior = potential_fn.prior
```

Comment:
I would prefer if this would not happen. In the future, we might have `potential_fn`s that do not have `self.prior`. In fact, the `EnsemblePotential` suggested in #612 will have the `self.prior` as a `List[torch.distribution]`. In that case, this line will fail.

The `self.prior` is only used in the warmup phase. I suggest to add an attribute `warm_up_target_dist` and use it here. If `warm_up_target_dist=None`, we just use `identity` as warm_up.

Reply:
Ah, I see yeah that makes sense. I currently also use the prior in the VIPosterior to check q and to infer the right transform if theta_transform is not given. So do we remove it there as well?

I would prefer to add an optional attribute "prior". If it is specified we use warm_up_target prior if not we use identitiy.

Reply:
That's also fine

Reply:
And yes, I'd remove it there as well.

Reply:
Mhh, I think a problem is that if theta_transform and the prior is not specified then one cannot determine the correct support of q. So I then just throw an error (that atleast on of them must be there), right?

Reply:
So I gues I also add an optional argument for the VIPosterior called "prior". (But then it also should be checked that the potential_fn "implicit" prior and the passed prior agrees).

Reply:
Yeah I guess one of them must be there.

And yes, you pass the prior to `VIPosterior`.

No need to check that the priors are the same though imo. We can't catch every little thing.

Reply:
As I said, please try to never access `potential.prior`. Not even for a "check" whether they are the same